_posts/2017/2017-02-02-programming-beyond-the-comfort-zone-javascript-elixir.md

```diff
@@ -0,0 +1,82 @@
+On top of that, it improves code readability - at least most of the time {% emoji wink %}.
+
+Now, what if I told you that you can learn all of this by using a web framework (and language) that is not only fun to work with but also faster and less error prone than all of the before mentioned?
+
+### Enter: Elixir and Phoenix
+
+This is rather a teaser that should make you curious about trying out something new and leaving your comfort zone, than a full blown tutorial or language reference.
+This first part will take a look at how **Elixir** might help you to rethink programming patterns leading to cleaner code and continue with what this **Phoenix** thing actually is.
+
+### The Platform
+
+Before we explore some Elixir code, let's start with a naive JS example of populating an array - mutation style.
+Don't fraud, I've seen a lot of people (including myself) starting like this (which is totally fine).
+
+```js
+const myNumbers = [-2, 3, 5, -34, 0, 32]
+@@ -38,7 +38,7 @@ const result = myNumbers
+ .map(num => num * 2)
+```
+
+This time we see that `result` will contain the double of all positive numbers from the source array.
+Way more explicit, isn't it?
+For the record, this is how the above would look in Elixir:
+
+```elixir
+my_numbers = [-2, 3, 5, -34, 0, 32]
+@@ -48,16 +48,16 @@ result = my_numbers
+
+```
+
+This pretty much resembles the JavaScript solution, right?!
+Almost!
+Since Elixir is based on immutable data, you could not have started like in the first JavaScript solution.
+You cannot simply mutate a data structure in Elixir!
+What you would have done, is to create a new list with every value from the old one that you see fits.
+Besides that, there are some other oddities.
+The `|>` is called a pipe operator and passes the value as the first argument to the next function.
+While we're using the `fn` keyword to define an anonymous function as the `filter` expression and explicitly naming the variable `num`, we're shortening this for the `map` function and just say: "*This function uses the first argument (`&1`) it receives and doubles it*".
+
+Let's take the example from above again.
+In contrast to JavaScript, we can use some sugar in Elixir called a `for` comprehension in conjunction with a filter to loop over the enumerable and create a new one (with all values > 0 doubled).
+
+```elixir
+my_numbers = [-2, 3, 5, -34, 0, 32]
+result = for n <- my_numbers, n > 0, do: n * 2
+```
+
+Note that the last statement of every Elixir function is automatically returned creating a new list bound to `result`.
+What actually looks like an assignment `=` is actually a `match`.
+Trying to match `result = 4` afterwards will give you an error, since the compiler tries to match `[6, 10, 64] = 4`.
+This might look strange at first, yet makes more sense once you grasp concepts like pattern matching or destructuring.
+
+Since we now have seen some language features, I have to bore you with some facts.
```

Comment:
Maybe better:
After looking at the above described language features, I cannot circumvent a few basic (and maybe boring {% emoji wink %}) facts.